finagle-base-http/src/main/scala/com/twitter/finagle/http/headers/MapBackedHeaderMap.scala

```diff
@@ -0,0 +1,108 @@
+import scala.collection.mutable
+
+/**
+ * Mutable, thread-safe [[HeaderMap]] implementation, backed by
+ * a mutable [[Map[String, Header]]], where the map key
+ * is forced to lower case
+ */
+final private[http] class MapBackedHeaderMap extends HeaderMap {
+  import HeaderMap._
+
+  // In general, HashSet/HashTables that are not thread safe are not
+  // durable to concurrent modification and can result in infinite loops.
+  // As such, we synchronize on the underlying `Headers` when performing
+  // accesses to avoid this. In the common case of no concurrent access,
+  // this should be cheap.
+  private[this] val underlying: mutable.Map[String, Header] = mutable.Map.empty
+
+  final override def foreach[U](f: ((String, String)) => U): Unit = for {
+    (_, h) <- underlying
+    kv <- h.iterator
+  } f((h.name, h.value))
+
+  // ---- HeaderMap -----
+
+  final def getAll(key: String): Seq[String] = underlying.synchronized {
+    underlying.get(key.toLowerCase).toSeq.flatMap(h => h.values)
+  }
+
+  // Validates key and value.
+  final def add(key: String, value: String): this.type = {
+    validateName(key)
+    addUnsafe(key, foldReplacingValidateValue(key, value))
+  }
+
+  // Does not validate key and value.
+  final def addUnsafe(key: String, value: String): this.type = underlying.synchronized {
+    val lower = key.toLowerCase
+    val header = new Header(key, value)
+    underlying.get(lower) match {
+      case Some(h) => h.add(header)
+      case None => underlying.+=((lower, header))
+    }
+    this
+  }
+
+  // Validates key and value.
+  final def set(key: String, value: String): this.type = {
+    validateName(key)
+    setUnsafe(key, foldReplacingValidateValue(key, value))
+  }
+
+  // Does not validate key and value.
+  final def setUnsafe(key: String, value: String): this.type = underlying.synchronized {
+    underlying.+=((key.toLowerCase, new Header(key, value)))
+    this
+  }
+
+  // ---- Map/MapLike -----
+
+  def -=(key: String): this.type = removed(key)
+  def +=(kv: (String, String)): this.type = set(kv._1, kv._2)
+
+
+  final def get(key: String): Option[String] = underlying.synchronized {
+    underlying.get(key.toLowerCase).map(_.value)
+  }
+
+  final def iterator: Iterator[(String, String)] = underlying.synchronized {
+    for {
+      (_, h) <- underlying.iterator
+      hx <- h.iterator
+    } yield (hx.name, hx.value)
+  }
+
+  final def removed(key: String): this.type = underlying.synchronized {
+    underlying.-=(key.toLowerCase)
+    this
+  }
+
+  final override def keys: Set[String] = keysIterator.toSet
+
+  final override def keysIterator: Iterator[String] = underlying.synchronized {
+    underlying.valuesIterator.flatMap(header => {
+      def rec(uniq: List[String], todo: List[String]): List[String] = todo match {
+        case Nil => uniq
+        case (head :: tail) =>
+          rec(head :: uniq, todo.filterNot(x => x == head))
+      }
+      rec(Nil, header.names.toList)
+    })
+  }
```

Comment:
I suspect this is costly both in terms of CPU and allocations. In the common case we will have only one entry in in the Headers linked list. Maybe we can make a hybrid implementation to the tune of
```scala
new Iterator[String] {
  private[this] val underlying = safeEntriesIterator() // the extracted array-backed iterator talked about in other comments
  private[this] var currentEntries: List[String] = Nil
  def hasNext = !currentEntries.isEmpty || underlying.hasNext
  def next(): String = {
    if (currentEntries.isEmpty) {
      val hs = underlying.next()
      if (hs.next == null) {
         // Take a shortcut for single value keys which is the common case
         hs.name
      } else {
        currentEntries = gatherEntries???(hs) // extract all unique and recurse one step since we know there will be currentEntries
        next()
      }
    } else {
       ??? // take one from currentEntries
    }
  }
}
```
It might be even better to use a second iterator instead of a `List` for `currentEntries`.